src/lock_free/mwcas_descriptor.cpp

```diff
@@ -0,0 +1,158 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// the corresponding header
+#include "dbgroup/atomic/mwcas/lock_free/mwcas_descriptor.hpp"
+
+// C++ standard libraries
+#include <atomic>
+#include <cstddef>
+
+// external libraries
+#include "dbgroup/lock/common.hpp"
+
+// local sources
+#include "dbgroup/atomic/mwcas/utility.hpp"
+
+namespace dbgroup::atomic::mwcas::lock_free
+{
+/*##############################################################################
+ * Static utilities
+ *############################################################################*/
+
+auto
+MwCASDescriptor::GetDescriptor()  //
+    -> MwCASDescriptor *
+{
+  /*auto *page = _gc->GetPageIfPossible<MwCASDescriptor>();
+  auto *desc = (page == nullptr) ? new MwCASDescriptor{} : static_cast<MwCASDescriptor *>(page);
+  desc->target_cnt_ = 0;*/
+  // GCを無視した一時的な実装
+  return new MwCASDescriptor{};
+}
+
+/*##############################################################################
+ * Utilities
+ *############################################################################*/
+auto
+MwCASDescriptor::MwCAS()  //
+    -> bool
+{
+  // set a memory fence
+  stat_.store(kUndecided, kRelease);
+  const auto succeeded = MwCASInternal();
+  return succeeded;
+}
+
+//                wordのビット割り当て（実装後に消す）
+// |     63     |       62-50       |      49-47     |        46-0       |
+// | MwCAS Flag | Reference Counter | Begin Position |Descriptor Address |
+
+auto
+MwCASDescriptor::MwCASInternal(  //
+    const size_t begin_pos       //
+    ) -> bool
+{
+  const auto desc_addr = std::bit_cast<uint64_t>(this) | kMwCASFlag;
+  auto cur_stat = stat_.load(kSeqCst);
+  if (cur_stat == kUndecided) {
+    auto stat = kSucceeded;
+    for (size_t i = begin_pos; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      if (!EmbedDescriptor(desc_addr, i, target.cnt)) {
+        stat = kFailed;
+        break;
+      }
+    }
+    cur_stat = stat_.load(kSeqCst);
+    if (cur_stat == kUndecided && stat_.compare_exchange_strong(cur_stat, stat, kSeqCst, kSeqCst)) {
+      cur_stat = stat;
+    }
+  }
+
+  auto ret = (cur_stat == kSucceeded);
+  auto target_cnt_sum = 0;
+  if (ret) {
+    for (size_t i = 0; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      auto expected = target.addr->load(kSeqCst);
+      if ((expected ^ desc_addr) & kAddrMask) continue;
+      target.addr->compare_exchange_strong(expected, target.new_val, target.fence, target.fence);
+      target_cnt_sum += target.cnt;
```

Comment:
`continue` のままだと記述子が交換済みの際にカウンタを参照できませんね．また，以下のsuggestionには含めていませんが，ここは他のスレッドが追従処理を開始している場合も考慮して実装してあげる必要があります．

```suggestion
      if (((expected ^ desc_addr) & kDescMask) == 0UL) {
        target.addr->compare_exchange_strong(expected, target.new_val, target.fence, target.fence);
      }
      target_cnt_sum += target.cnt.load(kSeqCst);
```

あと，ここ見て気づきましたが，ポインタとして抽出するための `kAddrMask` と別に，記述子が一致するかを判定するための `kDescMask = kMwCASFlag | kAddrMask` も必要です． `kAddrMask` のみだとユーザがたまたま同じ値をnew/old valueとして入れる可能性があります．

Reply:
`exit_cnt_`のインクリメントもアトミックにした方が良いでしょうか？

`exit_cnt_ > target_cnt_sum`には原理上ならないと思うので，GCすべきタイミングでGCされずに残り続けることはないと思いますが，すでに他のスレッドがGCした後にその場所をGCすることはあり得ると思います．もしそうなったらまずい気がするのですが，どうでしょう？

Reply:
はい、そこも `std::atomic` の利用が必須です。まず、 `std::atomic` を利用しない場合、加算をしてもコア間での値の共有が即座には行われません。各コアのCPUキャシュに加算した値が格納されて、適当なタイミングで同期が取られます。で、ここで問題となるのは、普通の加算はアトミックでないので合計が本来の値よりも小さくなる可能性があります（スレッドAとBがそれぞれ0から1に加算して、メモリ上のカウンタも1になってしまう）。